File: app/src/main/java/com/practice/forecast/ui/arch/mvvm/MvvmFragment.java
package com.practice.forecast.ui.arch.mvvm;

import android.content.Context;

import com.practice.forecast.ui.arch.Contract;

import java.lang.reflect.ParameterizedType;

import androidx.fragment.app.Fragment;

public class MvvmFragment<Host extends Contract.Host> extends Fragment {
    /**
     * the fragment callBack
     */
    private Host callBack;

    //@Override
    public final boolean hasCallBack() {
        return callBack != null;
    }

    public final boolean noHost() {
        return callBack == null;
    }

    /**
     * get the current fragment call back
     *
     * @return the current fragment call back
     */
    public final Host getCallBack() {
        return callBack;
    }

    @Override
    public void onAttach(Context context) {
        super.onAttach(context);
        // keep the call back
        try {
            this.callBack = (Host) context;
        } catch (Throwable e) {
            final String hostClassName = ((Class) ((ParameterizedType) getClass().
                    getGenericSuperclass())
                    .getActualTypeArguments()[1]).getCanonicalName();
            throw new RuntimeException("Activity must implement " + hostClassName
                    + " to attach " + this.getClass().getSimpleName(), e);
        }
    }

    @Override
    public void onDetach() {
        super.onDetach();
        // release the call back
        this.callBack = null;
    }
}

File: app/src/main/java/com/practice/forecast/ui/arch/mvvm/MvvmFragment.kt
package com.practice.forecast.ui.arch.mvvm

import android.content.Context
import androidx.fragment.app.Fragment
import com.practice.forecast.ui.arch.Contract
import java.lang.reflect.ParameterizedType

open class MvvmFragment<Host : Contract.Host?> : Fragment() {
    val newOne = ""
    val newOne1 = ""

    /**
     * get the current fragment call back
     *
     * @return the current fragment call back
     */
    /**
     * the fragment callBack
     */
    var callBack: Host? = null
        private set

    //@Override
    fun hasCallBack(): Boolean {
        return callBack != null
    }

    fun noHost(): Boolean {
        return callBack == null
    }

    override fun onAttach(context: Context) {
        super.onAttach(context)
        // keep the call back
        try {
            callBack = context as Host
        } catch (e: Throwable) {
            val hostClassName = ((javaClass.genericSuperclass as ParameterizedType?)
                    ?.getActualTypeArguments()?.get(1) as Class<*>).canonicalName
            throw RuntimeException("Activity must implement " + hostClassName
                    + " to attach " + this.javaClass.simpleName, e)
        }
    }

    override fun onDetach() {
        super.onDetach()
        // release the call back
        callBack = null
    }
}